From 511b3d965d07018d6f0eab8dc79fb707d493cebe Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Wed, 25 Feb 2015 16:50:07 -0800
Subject: [PATCH 082/184] MIPS: OCTEON: Add gpio_to_irq() test to
 octeon-error-injector.

Source: Cavium Networks, Inc.
MR: 00000
Type: Integration
Disposition: Merged from Octeon Tree
ChangeID: 9c9826259dac49cd9c06ec7983784585c535a685
Description:

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Corey Minyard <cminyard@mvista.com>
[Original patch taken from octeon-linux-kernel-patches-SDK-3.1.2-release]
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 arch/mips/cavium-octeon/octeon-error-injector.c | 34 +++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/arch/mips/cavium-octeon/octeon-error-injector.c b/arch/mips/cavium-octeon/octeon-error-injector.c
index 0ce0063..43f2791 100755
--- a/arch/mips/cavium-octeon/octeon-error-injector.c
+++ b/arch/mips/cavium-octeon/octeon-error-injector.c
@@ -11,6 +11,8 @@
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/slab.h>
+#include <linux/gpio.h>
+#include <linux/interrupt.h>
 
 #include <asm/octeon/octeon.h>
 #include <asm/octeon/cvmx-fpa.h>
@@ -110,6 +112,32 @@ static void octeon_error_injector_fpa1(void)
 	cvmx_write_csr(CVMX_FPA_POOLX_THRESHOLD(7), old_threshold);
 }
 
+static irqreturn_t octeon_error_injector_gpio_handler(int irq, void *arg)
+{
+	disable_irq_nosync(irq);
+	return IRQ_HANDLED;
+}
+
+static int octeon_error_injector_irq;
+
+static void octeon_error_injector_gpio_irq(void)
+{
+	int rv;
+	int irq = gpio_to_irq(test_param);
+
+	pr_err("gpio_to_irq(%d) -> %d\n", test_param, irq);
+
+	if (irq) {
+		rv = request_irq(irq, octeon_error_injector_gpio_handler, 0,
+				 "octeon_error_injector",
+				 octeon_error_injector_gpio_handler);
+		if (rv)
+			pr_err("request_irq failed: %d\n", rv);
+		else
+			octeon_error_injector_irq = irq;
+	}
+}
+
 static int __init octeon_error_injector_init(void)
 {
 	/* We are injecting errors, so mark the kernel as tainted.*/
@@ -131,6 +159,9 @@ static int __init octeon_error_injector_init(void)
 	case 5:
 		octeon_error_injector_l1_icache_parity();
 		break;
+	case 6:
+		octeon_error_injector_gpio_irq();
+		break;
 	default:
 		pr_err("Error: Unrecognized test number: %d\n",  test_number);
 		break;
@@ -142,6 +173,9 @@ module_init(octeon_error_injector_init);
 
 static void __exit octeon_error_injector_exit(void)
 {
+	if (octeon_error_injector_irq)
+		free_irq(octeon_error_injector_irq,
+			 octeon_error_injector_gpio_handler);
 }
 module_exit(octeon_error_injector_exit);
 
-- 
1.9.1

